guppylang/hresult.py

```diff
@@ -0,0 +1,218 @@
+r"""
+Quantinuum system results and utilities.
+
+Includes conversions to traditional distributions over bitstrings if a tagging
+convention is used, including conversion to a pytket BackendResult.
+
+Under this convention, tags are assumed to be a name of a bit register unless they fit
+the regex pattern `^([a-z][\w_]*)\[(\d+)\]$` (like `my_Reg[12]`) in which case they
+are assumed to refer to the nth element of a bit register.
+
+For results of the form ``` result("<register>", value) ``` `value` can be `{0, 1, True,
+False}`, wherein the register is assumed to be length 1, or lists over those values,
+wherein the list is taken to be the value of the entire register.
+
+For results of the form ``` result("<register>[n]", value) ``` `value` can only be `{0,
+1, True, False}`. The register is assumed to be at least `n+1` in size and unset
+elements are assumed to be `0`.
+
+Subsequent writes to the same register/element in the same shot will overwrite.
+
+To convert to a `BackendResult` all registers must be present in all shots, and register
+sizes cannot change between shots.
+
+"""
+
+from __future__ import annotations
+
+import re
+from collections import Counter, defaultdict
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+    from pytket.backends.backendresult import BackendResult
+
+#: Primitive data types that can be returned by a result
+DataPrimitive = int | float | bool
+#: Data value that can be returned by a result: a primitive or a list of primitives
+DataValue = DataPrimitive | list[DataPrimitive]
+TaggedResult = tuple[str, DataValue]
+# Pattern to match register index in tag, e.g. "reg[0]"
+REG_INDEX_PATTERN = re.compile(r"^([a-z][\w_]*)\[(\d+)\]$")
+
+
+@dataclass
+class HResult:
+    """Results from a single shot execution."""
+
+    entries: list[TaggedResult] = field(default_factory=list)
+
+    def __init__(self, entries: Iterable[TaggedResult] | None = None):
+        self.entries = list(entries or [])
+
+    def append(self, tag: str, data: DataValue) -> None:
+        self.entries.append((tag, data))
+
+    def as_dict(self) -> dict[str, DataValue]:
+        """Convert results to a dictionary.
+
+        For duplicate tags, the last value is used.
+
+        Returns:
+            dict: A dictionary where the keys are the tags and the
+            values are the data.
+
+        Example:
+            >>> results = Results()
+            >>> results.append("tag1", 1)
+            >>> results.append("tag2", 2)
+            >>> results.append("tag2", 3)
+            >>> results.as_dict()
+            {"tag1": 1, "tag2": 3}
+        """
+        return dict(self.entries)
+
+    def to_register_bits(self) -> dict[str, list[bool]]:
```

Comment:
I think, in its current state, this function should really be called `to_register_bools`. But I also wonder why you are converting everything to bools and then to bitstrings, it involves unnecessary back and forth conversion for the usual case where you are already given bits.

I'd think that `_cast_primitive_bool` should return a bit after doing necessary type checking and we stay away from bools for results.

Reply:
my initial handling of bools came about because a guppy `measure` returns a boolean, not an integer. This difference may be lost in the results handling (which is not too terrible because python booleans are integers anyway). I just wanted to avoid someone printing `result("c[0]", 3)` and accidentally getting that set to `1`, which we still can.

I agree with your simplification, and have made the suggested changes.